TD-2713 Enhancement - NumberField Uncontrolled

lukemojo opened the pull request:
<!-- Insert YouTrack link if relevant -->

Closes TD-2713

## Changes

Allow NumberField to work in an uncontrolled manner and allow WizardSteps to display an error without requiring a message.

This contains:

- Removal of internal NumberField validations
- Releasing of TextField props, omitting sx and type props
- Adding a new prop to wizard steps
- Convert files to typescript

A brief description of what this pull request solves / introduces.

- Removing min max validations and messages in NumberField and all props relating to it
  - Facilitates the removal of state from the component to avoid uncontrolled to controlled errors
- Remove all react state from NumberField
  - This allows the parent and our app validation handling to manage all error states and error messaging
- Update to allow TextField props in NumberField
  - This will allow more flexible usage in apps with less updates and releases needed for prop access in RUI
  - This aligns prop usage with MUI documentation
- Omit SX props and type from NumberField
  - Maintains current patterns
  - Ensure consistency of styling
  - Ensures NumberField cannot be used for other TextField purposes
- Add an error boolean in wizard steps
  -  This allows an error indicator to display without requiring a message or empty string
- Removes errorText prop
  -  Uses error the error boolean to style helperText with error styling
- Update helperText prop to only accept string

Please note that the field props are now in line with MUI TextField props and you should refer to the [MUI documentation](https://mui.com/material-ui/api/text-field/) for any props not specified in this components types. This change means that input attribute props are now accessed via inputProps as per the MUI documentation, e.g. `inputProps={{ step: x, min: x, max: x }}`. This is a change in behaviour and will need to be migrated when upgrading if used. Only input attributes are affected.

Additional work (should this be accepted) is to make similar change for releasing props on the other basic MUI components such as TextField, Checkbox etc.

## UI/UX

No change

## Testing notes

Please view the updated NumberField and WizardStep in storybook.

## Author checklist before assigning a reviewer

- [x] Reviewed my own code-diff.
- [x] Branch has been run in docker.
- [x] PR assigned to me or an appropriate delegate.
- [x] Relevant labels added to the PR.
- [x] Appropriate tests have been added.
- [x] Lint and test workflows pass.

dmbarry86 commented:
@lukemojo did you check through our codebase for any usage of this component that currently relies on the validation error handling which we'll need to now handle separately when we release this?

lukemojo commented:
> @lukemojo did you check through our codebase for any usage of this component that currently relies on the validation error handling which we'll need to now handle separately when we release this?

I did, there are no uses of min max and the validation error in Virto that I could find. The only usage is instrument designer. For that to be upgraded to latest RUI it would need to have its own validations.

dmbarry86 commented on src/Wizard/WizardSteps/WizardStep/WizardStep.tsx:
```diff
@@ -75,6 +78,9 @@ function WizardStepLabel({
     stepLabelProps.optional = errorText;
+  }
+
+  if (error || errorText) {
```
What was the background behind introducing error? If we're now relying on that to determine if we've got an error then shouldn't we only be checking for that rather than that or errorText? Presumably if error is false and errorText is set to some value then I shouldn't see the error?

lukemojo replied:
I introduced this because I was forced to set text to indicate the error but in the UI I'm displaying the error in the field/field group. I maintained the ability to use the error text to also set the error  as not to force a breaking change to how its currently being used. 

If we are making a breaking change I would suggest an error boolean and helperText to match how we display errors in inputs as we're replacing the helperText with errorText anyway. I ran a quick check and only BUILD is currently using the errorText in VIRTO

lukemojo replied:
I have updated as per discussion

dmbarry86 commented on src/Wizard/WizardSteps/WizardStep/WizardStep.types.ts:
```diff
@@ -5,6 +5,10 @@ export type WizardStepProps = {
+  /**
+   * If true, the step is marked as errored.
+   */
+  error?: boolean;
```
Missing default value for this prop in storybook docs. Whilst you are there could you add defaults for the others that are missing too? I assume default is false here for new error prop.

<img width="1090" alt="image" src="https://github.com/IPG-Automotive-UK/react-ui/assets/8976377/79b19d17-7f5e-458d-a3f2-1af6f393f8d4">

dmbarry86 reviewed (changes requested):
Changes look good. Just some missing defaults in storybook doc and a question over the logic to show error state.

dmbarry86 commented:
I think helper text should be showing in red for this story.

<img width="2214" alt="image" src="https://github.com/IPG-Automotive-UK/react-ui/assets/8976377/98ff930e-c650-4de5-a51a-177af584abad">

dmbarry86 reviewed (changes requested):
Fix WizardSteps error text story then we're good.

lukemojo commented:
> Fix WizardSteps error text story then we're good.

Thanks. This has now been updated.

dmbarry86 commented:
@lukemojo any reason this hasn't been merged yet? I notice you closed off the ticket in YouTrack. I think it probably should be merged to a v8 release branch so please align with @syedsalehinipg and get it merged today.

dmbarry86 commented:
@lukemojo any update here?

dmbarry86 commented:
@lukemojo what's the status on this one please? Can we either merge it to the appropriate release branch or delete it if no longer required?

lukemojo commented:
> @lukemojo what's the status on this one please? Can we either merge it to the appropriate release branch or delete it if no longer required?

It's still needed. I'll deal with this today. The work I'm currently doing is going on a minor version first.

lukemojo commented:
Apologies, this got into a bit of a mess. 

1.  I realised that I had used main as my intended branch when rebasing and merging and then pushed to remote, which caused a lot of conflicts. 
2. I realised the  release branch and main had diverged after incorporating main. 
3. Main has some problems and is not ready for a release so isn't currently being merged into the release/v8.0.0 branch until it is fixed.

To fix my branch, I checked the changes against main, I returned the branch to my final commit change that was originally reviewed, prior to the merge/rebase mistake. I've now incorporated and moved the base back to the release branch and force pushed the changes over the incorrect remote. 

Unfortunately, all of the shenanigans have dismissed your review @dmbarry86, apologies. Nothing has changed in the code, nothing additional has been added, I've just returned it to a mergeable state.
